Reader.hpp

```diff
@@ -15,26 +20,30 @@ namespace mae
 class EXPORT_MAEPARSER Reader
 {
   private:
-    MaeParser* m_mae_parser;
+    std::shared_ptr<MaeParser> m_mae_parser;
+#ifdef ZLIB_AVAILABLE
+    std::shared_ptr<std::ifstream> m_pregzip_stream;
+    std::shared_ptr<boost::iostreams::filtering_streambuf<boost::iostreams::input> > m_gzip_stream;
+#endif
 
   public:
     Reader(FILE* file, size_t buffer_size = BufferLoader::DEFAULT_SIZE)
```

Comment:
My bias towards switching to a smart pointer is that I made the mistake of passing a local variable in by reference and didn't even think about it, and when this happens it's *really* hard to debug, because the crash doesn't even happen until you've read a bunch more structures, and it's unclear why it's crashing.  This protects the user from making that not-crazy and very-hard-to-debug mistake.